activation/poet.go

```diff
@@ -257,13 +262,15 @@ func (c *HTTPPoetClient) req(ctx context.Context, method, path string, reqBody,
 	if res.StatusCode != http.StatusOK {
 		c.logger.Info("got poet response != 200 OK", zap.String("status", res.Status), zap.String("body", string(data)))
 	}
-	switch res.StatusCode {
-	case http.StatusOK:
-	case http.StatusNotFound:
+	switch {
+	case res.StatusCode == http.StatusOK:
+	case res.StatusCode == http.StatusNotFound:
 		return fmt.Errorf("%w: response status code: %s, body: %s", ErrNotFound, res.Status, string(data))
-	case http.StatusServiceUnavailable:
+	case res.StatusCode == http.StatusServiceUnavailable:
 		return fmt.Errorf("%w: response status code: %s, body: %s", ErrUnavailable, res.Status, string(data))
-	case http.StatusBadRequest:
+	case res.StatusCode == http.StatusBadRequest && strings.Contains(string(data), registration.ErrTooLateToRegister.Error()):
+		return fmt.Errorf("%w: response status code: %s, body: %s", ErrSubmitTooLate, res.Status, string(data))
```

Comment:
I'm not a huge fan of asserting strings in the response. Arguably `ErrSubmitTooLate` doesn't need to be its own error since it is an invalid request independent of why. `ErrSubmitTooLate` is also only used in a test where asserting that the returned error is `ErrInvalidRequest` in the case of a deadline that cannot be met would in my opinion be sufficient as well.

The full error text still contains the info that it was because of the deadline for additional information in logging.

Reply:
Me neither, but GRPC doesn't really give another way (at least I'm not aware of it) to signal different error reasons.

I need to distinguish between just an invalid request (for example bad pow params, signature etc. that are bugs) and a plausible scenario of submitting too late. This latter is handled in the code by expiring the ATX.

Reply:
What I mean is all of these (bad pow params, signature and deadline) are invalid requests in the case of the first two the node just tries again 5 minutes later (probably receiving the same response) but with the later it expires the challenge.

Shouldn't it expire the challenge either way?